src/OrchardCore/OrchardCore.Infrastructure/Shell/ShellDescriptorManager.cs

```diff
@@ -98,32 +94,26 @@ public async Task UpdateShellDescriptorAsync(int priorSerialNumber, IEnumerable<
                 _logger.LogInformation("Updating shell descriptor for tenant '{TenantName}' ...", _shellSettings.Name);
             }
 
-            if (shellDescriptorRecord == null)
-            {
-                shellDescriptorRecord = new ShellDescriptor { SerialNumber = 1 };
-            }
-            else
-            {
-                shellDescriptorRecord.SerialNumber++;
-            }
-
-            shellDescriptorRecord.Features = _alwaysEnabledFeatures.Union(enabledFeatures).ToList();
-            shellDescriptorRecord.Installed = shellDescriptorRecord.Installed.Union(shellDescriptorRecord.Features).ToList();
+            shellDescriptor.SerialNumber++;
+            shellDescriptor.Features = _alwaysEnabledFeatures.Union(enabledFeatures).ToList();
+            shellDescriptor.Installed = shellDescriptor.Installed.Union(shellDescriptor.Features).ToList();
 
             if (_logger.IsEnabled(LogLevel.Information))
             {
                 _logger.LogInformation("Shell descriptor updated for tenant '{TenantName}'.", _shellSettings.Name);
             }
 
-            _session.Save(shellDescriptorRecord);
+            await _documentStore.UpdateAsync(shellDescriptor, _ => Task.CompletedTask);
 
             // In the 'ChangedAsync()' event the shell will be released so that, on request, a new one will be built.
             // So, we commit the session earlier to prevent a new shell from being built from an outdated descriptor.
-
-            await _session.SaveChangesAsync();
```

Comment:
So here, keep this line but just change it to.

    await _documentStore.CommitAsync();

I did some tests, as I have tried it works well.

The only thing depending on what you plan to do is that currently `_documentStore.CommitAsync()` throws an exception if it is the `FileDocumentStore`, but easy to change.

Reply:
I opened another PR (#12942) to correct `FileDocumentStore` commit behavior.

Reply:
Okay will look at your other PR asap.

Yes from memory currently the file document store was intentionally kept a little coupled to the ambient DocumentStore/ISession, so that a given file is only updated after a commit success meaning that the default document store was not cancelled after some validations in other places.

The ShellDescriptor is a special case, in any case we still want to persist it immediately.